common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java

```diff
@@ -106,28 +127,32 @@ public MaterialConfig config() {
     }
 
     public List<Modification> modificationsSince(File baseDir, Revision revision, final SubprocessExecutionContext execCtx) {
+        unshallowIfNeeded(ProcessOutputStreamConsumer.inMemoryConsumer(), revision, baseDir);
         GitCommand gitCommand = getGit(baseDir);
         return gitCommand.modificationsSince(revision);
     }
 
     public MaterialInstance createMaterialInstance() {
-        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString());
+        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString(), shallowClone);
     }
 
     @Override
     protected void appendCriteria(Map<String, Object> parameters) {
         parameters.put(ScmMaterialConfig.URL, url.forCommandline());
         parameters.put("branch", branch);
+        parameters.put("shallowClone", shallowClone);
```

Comment:
@arvindsv and @ketan : I have added ShallowClone flag into the material criteria -- which makes material with different ShallowClone config stored separately in DB. I feel this is important so that different pipeline can use different clone settings, e.g using shallow clone in test pipeline and using full clone in heroku deploy pipeline. 

There is one implications to this change: this will change how git materials generate fingerPrint. One thing that worries me is old materials in the db will have a fingerPrint value that out of sync. I manually tested the upgrading process, seems like go server do not really cares about that.  But my knowledge on go is too limited, so want to ask you guys is this fine?